index.js

```diff
@@ -202,9 +186,28 @@ Choo.prototype.toString = function (location, state) {
   assert.equal(typeof location, 'string', 'choo.toString: location should be type string')
   assert.equal(typeof this.state, 'object', 'choo.toString: state should be type object')
 
-  this.state.href = location.replace(/\?.+$/, '')
-  this.state.query = nanoquery(location)
-  var html = this.router(location)
+  var href = location.replace(/\?.+$/, '')
+  this._matchRoute(href, location)
```

Comment:
We could move this into the `pushState`/`popState`/`replaceState` methods.
But there might be other side effects if we do that.

Reply:
Perhaps we should leave this the way it is for this PR, and consider the changes in a seperate PR? That might keep the changes nicely contained, and easier to follow along.

Reply:
Yes, I completely agree here. This definitely needs more elaboration.

Reply:
The question here was whether we should prevent `_matchRoute` from mutating the state. It should only return the new params and mutate the state after the `NAVIGATE` event got triggered.
This would allow us to keep the current state but already pass route params to some function to preload something.

This could end up in something similar to the angular `resolve` object where it's possible to preload some resources based on route params. In there we could also cancel the route change if a `resolve` gets rejected.
Maybe this would make everything more complicated. I'll think about it and will create a new issue if I think it's a good idea in the longer term. Here's a link to the resolve reference documentation in case you don't know it: 
https://angular.io/api/router/Resolve

And a few more readable examples:
https://www.codementor.io/max_diachenko/angular-2-route-resolves--typescript-du1089pdm
https://medium.com/@dazcyril/resolving-data-in-angular-2-4-and-5-refactoring-components-and-moving-to-ngrx-store-15e15d8be3dd#9968

Reply:
I'm really eager to get this released! :D
Without this it's hard to redirect to other routes based on where the user is trying to go (prevent user from going to routes they should not have access to).